### ISSUE TITLE
content: "ARIA: poison or antidote?" vs semantic elements

### ISSUE THREAD

**Issue by tehoko** [GoogleChrome/web.dev#3757]
### What page needs to be updated?

- https://web.dev/aria-poison-or-antidote/ by @aleventhal (Aaron's last name is misspelled in the article, by the way)
- https://github.com/GoogleChrome/web.dev/tree/master/src/site/content/en/blog/aria-poison-or-antidote by @kaycebasques 

### Why is this update needed?

I feel the article propagates an idea that HTML **needs** additional `aria-*` attributes to be accessible. Meanwhile, proper use of [semantic elements](https://www.w3schools.com/html/html5_semantic_elements.asp) should make the page accessible out of the box, without as much additional work as the article implies.

Perhaps I'm missing the point of the article – given it was written by the developer of the first implementation of ARIA in a browser and who knows more than me – but it doesn't hurt to ask.

### Examples

> Menu bars are so useful you would think they would be part of HTML, but they aren't.

This one is puzzling to me, especially since Addendum 2 mentions semantic elements ("this container over here is a navigation panel") – in this case, the `<nav>` element, which is for the main navigation. Doesn't that fit "menu bar", which the article says HTML is lacking?

---

> Let's make a menu bar together. We show a bunch of items in generic box elements called divs.

`<div>`s are not proper elements for this. This is the core problem with the article: the author suggests using generic box elements, which then must be sprinkled with `aria-*`, `event.preventDefault()`, and other spices. All of that to emulate what a `<button>` or `<a>` element would do out of the box.

---

> Our menu bar was created with duct tape and divs. As a result, a screen reader has no idea what any of it is.

Indeed. But if it was a `<nav>` with an unordered list of `<button>`s or `<a>`s, the screen reader would parse it without any additional markup needed.

```html
<nav>
  <ul>
    <li>(menu item)</li>
    <li>(menu item)</li>
    <li>
      <ul>
        (nested submenu)
      </ul>
    </li>
  </ul>
</nav>
```

Such semantic approach is also easier to style with CSS. It will even work on text-based web browsers or if stylesheets are disabled.

---

> Here are some common uses of ARIA.
> \- Special widgets that don't exist in HTML, like a **menu bar**, autocomplete, **tree**, or spreadsheet

As mentioned above, menu bar can be implemented with `<nav>` instead of nested `<div>`s with `aria-*` attributes that say "I'm a navigation element".

Tree (assuming it's about something like exploring a directory of files and folders) can be done with `<details>` and `<summary>` elements. The resulting tree can be expanded and collapsed with no additional JavaScript code.

```html
<details>
  <summary>Cute animals</summary>
  <details>
    <summary>Dogs</summary>
    <p>Woof.jpg</p>
    <p>Bark.jpg</p>
  </details>
  <details>
    <summary>Cats</summary>
    <p>Meow.jpg</p>
    <p>Hiss.tiff</p>
  </details>
</details>
<details>
  <summary>Wallpapers</summary>
  <p>Vector.svg</p>
  <p>Outline.png</p>
</details>
```

**Comment by tehoko** [comment]
It's unfortunate that this issue is now marked as stale.

I feel leaving this article unchanged does a disservice to the web developer community. Instead of teaching good practices on how to use dedicated semantic elements, it suggests applying a band-aid on dated markup.

The author's name is spelled correctly now, so there's that.

**Comment by aleventhal** [comment]
@tehoko , I just saw this now. Thanks for the feedback. I'd love to update the article to be more accurate. I do believe the point of the article, which is to explain why ARIA needs such thorough testing and breaks so easily, is still a vital message.

Menubar: It looks like you are indeed correct that a newer up-to-date practice is to use nav, ul and li instead of a div. However, ARIA is still needed. Take a look at the ARIA practices example for a menubar: https://w3c.github.io/aria-practices/examples/menubar/menubar-navigation.html

For the tree: one can also use more semantic markup underneath. Looking at the current ARIA practices recommendation, it uses nested lists, not details/summary. I'm not sure why. But, you'll still need tree/treeitem to get the proper screen reader UX -- as a user navigates to a tree item, they hear that it's a "tree item", letting them know what left/right/up/down will do. They will hear the level, position in group, and group size, and whether the item is expanded. Try  https://w3c.github.io/aria-practices/examples/treeview/treeview-1/treeview-1a.html with a screen reader vs. your details/summary example

In either approach, without the ARIA, a screen reader user will not hear when they are on a menu item, when that menu item is expandable, etc.  Your suggestions might kind of prove the point. I think that HTML authors might think that the semantic markup in HTML is enough to provide the screen reader experience a user will understand. The semantics you're using are prettier code-wise, or easier to style, but they don't work as well with a screen reader. That's the issue :/

Interested in hearing your thoughts.

**Comment by tehoko** [comment]
Thanks for the reply!

Using NVDA Screen Reader, I've checked a website that used a very basic flat menu build with `<nav>` and `<li>` items. It correctly described entering "navigation landmark" when focused, so at least this part can be considered covered. Changing `<div>` to `<nav>` can definitely help. I still have to try building a more complex, nested menu, and see how the reader would interpret it, compared to one relying on ARIA attributes.

With the same NVDA reader, I've looked into collapsible groups that use details/summary. I thought you meant that screen readers wouldn't even say that such a group is collapsed or expanded; thankfully, it properly read the group's state, so at least this information is not lost. There's no information that it's a tree item, though, so `aria` is ultimately needed, as you said.

Note that the tree example requires quite a bit of JS code to handle all the navigation features, while details/summary supports tabbing and expanding (enter or space) out of the box. Maybe a combination of both approaches could be considered.

I share your disappointment that semantic elements don't provide proper accessibility yet, but I feel their potential should not be overlooked. If we can use a more fitting element, we should use it instead of wrapping everything in a `<div>`.
